### PR TITLE
Add SPDM 1.3 new feature: Set key pair info

### PR DESCRIPTION
**Pull request by Wenxing-hou**
The PR is pure set_key_pair_info feature now.

### CODE REVIEW

File: include/hal/library/responder/key_pair_info.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * @param  key_pair_id                  Indicate which key pair ID's information to retrieve.
+ * @param  operation                    Set key pair info operation: change/erase/generate.
+ * @param  desired_key_usage            Indicate the desired key usage for the requested key pair ID.
+ * @param  desired_asym_algo            Indicate the desired asymmetric algorithm for the requested key pair ID.
+ * @param  desired_assoc_cert_slot_mask Indicate the desired certificate slot association for the requested key pair ID.
+ *
+ * @retval true  set key pair info successfully.
+ * @retval false set key pair info failed.
+ **/
+extern bool libspdm_write_key_pair_info(
```
Comment (steven-bellock) [inline include/hal/library/responder/key_pair_info.h:65]: As discussed in https://github.com/DMTF/SPDM-WG/issues/3535, Responder may need to be reset for some of these operations to take place, so add a `bool *need_reset` parameter.

**Review by steven-bellock — CHANGES_REQUESTED** [review]

**Comment by steven-bellock** [comment]
Filed https://github.com/DMTF/SPDM-WG/issues/3588.

File: library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    if (!libspdm_is_capabilities_flag_supported(
+            spdm_context, false, 0,
+            SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_SET_KEY_PAIR_INFO_CAP)) {
+        return libspdm_generate_error_response(
+            spdm_context, SPDM_ERROR_CODE_UNSUPPORTED_REQUEST,
+            SPDM_SET_KEY_PAIR_INFO, response_size, response);
+    }
+
+    key_pair_id = spdm_request->key_pair_id;
+    if (key_pair_id == 0) {
```
[the remark this answers is not in the document]
Comment (Wenxing-hou) [reply library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c]: Thanks. I have added the check.

File: include/hal/library/responder/key_pair_info.h (hunk 1 above)
Comment (Wenxing-hou) [reply include/hal/library/responder/key_pair_info.h:69]: Yes. I have added the `need_reset` parameter and added the unit_test for the reset case.

File: include/hal/library/responder/key_pair_info.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ **/
+
+#ifndef RESPONDER_KEY_PAIR_INFO_H
+#define RESPONDER_KEY_PAIR_INFO_H
+
+#include "hal/base.h"
+#include "internal/libspdm_lib_config.h"
+#include "industry_standard/spdm.h"
+
+#if (LIBSPDM_ENABLE_CAPABILITY_GET_KEY_PAIR_INFO_CAP || \
```
[the remark this answers is not in the document]
Comment (Wenxing-hou) [reply include/hal/library/responder/key_pair_info.h:14]: But the libspdm_rsp_set_key_pair_info_ack will call the `libspdm_read_key_pair_info` to get the capbilities. Thanks.

[the remark this answers is not in the document]
Comment (steven-bellock) [reply include/hal/library/responder/key_pair_info.h:14]: `GET_KEY_PAIR_INFO_CAP` is required to be set if `SET_KEY_PAIR_INFO_CAP` is set. If `SET_KEY_PAIR_INFO_CAP` is set but `GET_KEY_PAIR_INFO_CAP` isn't, then it would result in a compilation error and that is fine.

File: library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+            session_info->secured_message_context);
+        if (session_state != LIBSPDM_SESSION_STATE_ESTABLISHED) {
+            return libspdm_generate_error_response(
+                spdm_context,
+                SPDM_ERROR_CODE_UNEXPECTED_REQUEST, 0,
+                response_size, response);
+        }
+    }
+
+    if (!libspdm_is_capabilities_flag_supported(
```
Comment (steven-bellock) [inline library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c:98]: This capability is not related to setting keys. As such libspdm should not use or check it in 1.3. Correct @jyao1?

File: library/spdm_requester_lib/libspdm_req_set_key_pair_info.c [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    if (operation > SPDM_SET_KEY_PAIR_INFO_GENERATE_OPERATION) {
+        return LIBSPDM_STATUS_INVALID_PARAMETER;
+    }
+    if (operation == SPDM_SET_KEY_PAIR_INFO_ERASE_OPERATION) {
+        if ((desired_key_usage != 0) || (desired_asym_algo != 0) ||
+            (desired_assoc_cert_slot_mask != 0)) {
+            return LIBSPDM_STATUS_INVALID_PARAMETER;
+        }
+    }
+    if (operation == SPDM_SET_KEY_PAIR_INFO_GENERATE_OPERATION) {
```
Comment (steven-bellock) [inline library/spdm_requester_lib/libspdm_req_set_key_pair_info.c:56]: This check is not needed. The Requester can ask the Responder to generate a key without having `DesiredKeyUsage`, `DesiredAsymAlgo`, or `DesiredAssocCertSlotMask` set. The Requester can later use the `ParameterChange` operation to set these.

**Review by steven-bellock — CHANGES_REQUESTED** [review]

File: include/hal/library/responder/key_pair_info.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyao1) [reply include/hal/library/responder/key_pair_info.h:14]: agree.

File: library/spdm_requester_lib/libspdm_req_set_key_pair_info.c (hunk 5 above)
Comment (jyao1) [reply library/spdm_requester_lib/libspdm_req_set_key_pair_info.c]: agree.

File: library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c (hunk 4 above)
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c:98]: I dont see anything wrong with this check.

Comment (steven-bellock) [reply library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c:98]: @Wenxing-hou already addressed it.

**Review by steven-bellock — CHANGES_REQUESTED** [review]

**Comment by jyao1** [comment]
@steven-bellock , could you please give all comment one time, instead of one by one?

**Comment by steven-bellock** [comment]
@jyao1 I am done with the review.

File: library/spdm_requester_lib/libspdm_req_set_key_pair_info.c [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    context->crypto_request = true;
+    retry = context->retry_times;
+    retry_delay_time = context->retry_delay_time;
+    do {
+        status = libspdm_try_set_key_pair_info(context, session_id, key_pair_id,
+                                               operation,
+                                               desired_key_usage,
+                                               desired_asym_algo,
+                                               desired_assoc_cert_slot_mask);
+        if ((status != LIBSPDM_STATUS_BUSY_PEER) || (retry == 0)) {
```
Comment (steven-bellock) [inline library/spdm_requester_lib/libspdm_req_set_key_pair_info.c:203]: Don't need the `retry == 0` anymore. See https://github.com/DMTF/libspdm/pull/2815.

Comment (jyao1) [reply library/spdm_requester_lib/libspdm_req_set_key_pair_info.c:203]: Good catch. Fixed.

File: library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+            SPDM_SET_KEY_PAIR_INFO, response_size, response);
+    }
+
+    if (operation > SPDM_SET_KEY_PAIR_INFO_GENERATE_OPERATION) {
+        return libspdm_generate_error_response(spdm_context,
+                                               SPDM_ERROR_CODE_INVALID_REQUEST, 0,
+                                               response_size, response);
+    }
+
+    if (operation == SPDM_SET_KEY_PAIR_INFO_ERASE_OPERATION) {
```
[the remark this answers is not in the document]
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c:216]: ok. combined.

File: library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+                                               SPDM_ERROR_CODE_INVALID_REQUEST, 0,
+                                               response_size, response);
+    }
+    if (((capabilities & SPDM_KEY_PAIR_CAP_ERASABLE_CAP) == 0) &&
+        (operation == SPDM_SET_KEY_PAIR_INFO_ERASE_OPERATION)) {
+        return libspdm_generate_error_response(spdm_context,
+                                               SPDM_ERROR_CODE_INVALID_REQUEST, 0,
+                                               response_size, response);
+    }
+    if (((capabilities & SPDM_KEY_PAIR_CAP_CERT_ASSOC_CAP) == 0) &&
```
[the remark this answers is not in the document]
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c:172]: Add check for desired_assoc_cert_slot_mask != 0

File: library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        &capabilities,
+        &key_usage_capabilities,
+        &current_key_usage,
+        &asym_algo_capabilities,
+        &current_asym_algo,
+        &assoc_cert_slot_mask,
+        NULL, NULL);
+    if (!result) {
+        return libspdm_generate_error_response(spdm_context,
+                                               SPDM_ERROR_CODE_INVALID_REQUEST, 0,
```
[the remark this answers is not in the document]
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c:130]: agree. changed.

File: library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    result = libspdm_read_key_pair_info(
+        spdm_context,
+        key_pair_id,
+        &capabilities,
+        &key_usage_capabilities,
+        &current_key_usage,
+        &asym_algo_capabilities,
+        &current_asym_algo,
+        &assoc_cert_slot_mask,
+        NULL, NULL);
```
[the remark this answers is not in the document]
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c:127]: comment is added.

File: library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+        return libspdm_generate_error_response(spdm_context,
+                                               SPDM_ERROR_CODE_OPERATION_FAILED, 0,
+                                               response_size, response);
+    }
+
+    spdm_response = response;
+    *response_size = sizeof(spdm_set_key_pair_info_ack_response_t);
+
+
+    if (libspdm_is_capabilities_flag_supported(
```
[the remark this answers is not in the document]
Comment (jyao1) [reply library/spdm_responder_lib/libspdm_rsp_set_key_pair_info_ack.c:253]: yes

**Review by steven-bellock — APPROVED** [review]
Is @Xiaohanjlll going to write some unit tests? Or should I do that?
